pkg/controller/statefulset/stateful_pod_control.go

```diff
@@ -108,23 +157,44 @@ func (spc *realStatefulPodControl) UpdateStatefulPod(set *apps.StatefulSet, pod
 				return err
 			}
 		}
+		if utilfeature.DefaultFeatureGate.Enabled(features.StatefulSetAutoDeletePVC) {
+			// if the Pod's PVCs are not consistent with the StatefulSet's PVC deletion policy, update the PVC
+			// and dirty the pod.
+			if match, err := spc.ClaimsMatchRetentionPolicy(set, pod); err != nil {
+				if consistent {
+					// Oops, something has gone badly wrong.
+					spc.recordPodEvent("update", set, pod, err)
+					return err
+				}
+				// Otherwise the error may be due to some other update, and we'll have to reconcile the next time around.
+				// TODO: does the reconcile need to be sped up?
+			} else if !match {
+				if err := spc.UpdatePodClaimForRetentionPolicy(set, pod); err != nil {
+					spc.recordPodEvent("update", set, pod, err)
+					return err
+				}
+				consistent = false
+			}
+		}
+
 		// if the Pod is not dirty, do nothing
 		if consistent {
 			return nil
 		}
 
 		attemptedUpdate = true
 		// commit the update, retrying on conflicts
-		_, updateErr := spc.client.CoreV1().Pods(set.Namespace).Update(context.TODO(), pod, metav1.UpdateOptions{})
+
+		updateErr := spc.objectMgr.UpdatePod(pod)
 		if updateErr == nil {
 			return nil
 		}
 
-		if updated, err := spc.podLister.Pods(set.Namespace).Get(pod.Name); err == nil {
+		if updated, err := spc.objectMgr.GetPod(set.Namespace, pod.Name); err == nil {
```

Comment:
Why is this a new change?

Reply:
Why are we doing a live read here?  UpdatePod should return the resulting pod object?

Reply:
So I think a problem here is that you're acting based on the cache, then updating the object, then bypassing the cache *just for this call*, but the the next call is going to go back to the cache.  If the ONLY thing you're doing is printing the pod, then you should get it from the UpdatePod call, not by doing a live read, or viewing the cache.

Reply:
This was pre-existing code, so I'm not sure what the intention was. It looks like this get is only done if UpdatePod returns an error, so presumably the pod returned from that call may not be valid?

Reply:
Ok, objectMgr is doing *some* calls with the lister, and some calls without.  As long as all reads are from the lister, then this is ok.  I missed that, thanks for pointing it out.